nf_core/modules.py

```diff
@@ -231,3 +237,481 @@ def has_valid_pipeline(self):
+
+        For all nf-core modules, the correct file structure is assured and important
+        file content is verified. If directory subject to linting is a clone of 'nf-core/modules',
+        the files necessary for testing the modules are also inspected.
+
+        For all local modules, the '.nf' file is checked for some important flags, and warnings
+        are issued if untypical content is found.
+
+        :param module:          A specific module to lint
+        :param print_results:   Whether to print the linting results
+        :param show_passed:     Whether passed tests should be shown as well
+
+        :returns:               dict of {passed, warned, failed}
+        """
+
+        # Get list of all modules in a pipeline
+        local_modules, nfcore_modules = self.get_installed_modules()
+
+        # Only lint the given module
+        # TODO --> decide whether to implement this for local modules as well
+        if module:
+            local_modules = []
+            nfcore_modules_names = [m.split("software" + os.sep)[1] for m in nfcore_modules]
+            try:
+                idx = nfcore_modules_names.index(module)
+                nfcore_modules = [nfcore_modules[idx]]
+            except ValueError as e:
+                log.error("Could not find the given module!")
+                sys.exit(1)
+
+        # Lint local modules
+        self.lint_local_modules(local_modules)
+
+        # Lint nf-core modules
+        self.lint_nfcore_modules(nfcore_modules)
+
+        if print_results:
+            self._print_results(show_passed=show_passed)
+
+        return {"passed": self.passed, "warned": self.warned, "failed": self.failed}
+
+    def lint_local_modules(self, local_modules):
+        """
+        Lint a local module
+        Only issues warnings instead of failures
+        """
+        for mod in local_modules:
+            self.lint_main_nf(mod)
+
+    def lint_nfcore_modules(self, nfcore_modules):
+        """
+        Lint nf-core modules
+        For each nf-core module, checks for existence of the files
+        - main.nf
+        - meta.yml
+        - functions.nf
+        And verifies that their content.
+
+        If the linting is run for modules in the central nf-core/modules repo
+        (repo_type==modules), files that are relevant for module testing are
+        also examined
+        """
+        # Iterate over modules and run all checks on them
+        for mod in nfcore_modules:
+            if "TOOL/SUBTOOL" in mod:
+                continue
+            module_name = mod.split(os.sep)[-1]
+
+            # Lint the main.nf file
+            inputs, outputs = self.lint_main_nf(os.path.join(mod, "main.nf"))
+
+            # Lint the meta.yml file
+            self.lint_meta_yml(os.path.join(mod, "meta.yml"), module_name, inputs=inputs, outputs=outputs)
+
+            # Lint the functions.nf file
+            self.lint_functions_nf(os.path.join(mod, "functions.nf"))
```

Comment:
I wonder if it's worth making another class for all this stuff, eg `NFCoreModule` or something. Can then initialise it with the path and have each of the linting functions contained with that class with all of the module-specific data contained there. Might be a bit cleaner and easier to maintain / reuse.

See `nf-core list` for an example of where I have done this when listing local and remote workflows. That code ended up being reused all over the place so was quite lucky / the result of a fair bit of code refactoring 😉

Reply:
Hmm yes I like that idea 🤔 will probably do some refactoring today